File: scripts/clean-memory-garden-hotspots.js
/** Remove baked UI hotspot rectangles from memory garden room art. */
const fs = require("fs");
const path = require("path");
const sharp = require("sharp");

const root = path.resolve(__dirname, "..");
const src = path.join(root, "assets", "locus-chamber-memory-garden-room.png");
const bak = path.join(root, "assets", "locus-chamber-memory-garden-room.source.png");

function luminance(r, g, b) {
  return 0.299 * r + 0.587 * g + 0.114 * b;
}

function isOverlayPixel(r, g, b) {
  if (Math.max(r, g, b) - Math.min(r, g, b) > 55) return false;
  const lum = luminance(r, g, b);
  if (lum < 155) return false;
  return r > 145 && g > 135 && b > 115 && lum < 245;
}

async function main() {
  if (!fs.existsSync(src)) {
    console.error("Missing", src);
    process.exit(1);
  }

  if (!fs.existsSync(bak)) {
    fs.copyFileSync(src, bak);
  }

  const { data, info } = await sharp(bak).removeAlpha().raw().toBuffer({ resolveWithObject: true });
  const { width: w, height: h, channels } = info;
  const mask = new Uint8Array(w * h);

  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const i = (y * w + x) * channels;
      if (isOverlayPixel(data[i], data[i + 1], data[i + 2])) {
        mask[y * w + x] = 1;
      }
    }
  }

  const dilated = new Uint8Array(w * h);
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (!mask[y * w + x]) continue;
      for (let dy = -2; dy <= 2; dy++) {
        for (let dx = -2; dx <= 2; dx++) {
          const ny = y + dy;
          const nx = x + dx;
          if (ny >= 0 && ny < h && nx >= 0 && nx < w) {
            dilated[ny * w + nx] = 1;
          }
        }
      }
    }
  }

  let changed = 0;
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (!dilated[y * w + x]) continue;
      const samples = [];
      for (let dy = -4; dy <= 4; dy++) {
        for (let dx = -4; dx <= 4; dx++) {
          if (dx === 0 && dy === 0) continue;
          const ny = y + dy;
          const nx = x + dx;
          if (ny < 0 || ny >= h || nx < 0 || nx >= w) continue;
          if (dilated[ny * w + nx]) continue;
          const si = (ny * w + nx) * channels;
          samples.push([data[si], data[si + 1], data[si + 2]]);
        }
      }
      if (!samples.length) continue;
      const i = (y * w + x) * channels;
      data[i] = Math.round(samples.reduce((s, c) => s + c[0], 0) / samples.length);
      data[i + 1] = Math.round(samples.reduce((s, c) => s + c[1], 0) / samples.length);
      data[i + 2] = Math.round(samples.reduce((s, c) => s + c[2], 0) / samples.length);
      changed++;
    }
  }

  await sharp(data, { raw: { width: w, height: h, channels } })
    .blur(0.35)
    .png({ compressionLevel: 9 })
    .toFile(src);

  console.log("Cleaned", changed, "overlay pixels in", src);
}

main().catch(function (err) {
  console.error(err);
  process.exit(1);
});

File: scripts/clean-memory-garden-hotspots.py
"""Remove baked UI hotspot rectangles from memory garden room art."""
from __future__ import annotations

import shutil
from pathlib import Path

try:
    from PIL import Image, ImageFilter
except ImportError:
    raise SystemExit("Pillow required: pip install pillow")

ROOT = Path(__file__).resolve().parents[1]
SRC = ROOT / "assets" / "locus-chamber-memory-garden-room.png"
BAK = ROOT / "assets" / "locus-chamber-memory-garden-room.source.png"
OUT = SRC


def luminance(r: int, g: int, b: int) -> float:
    return 0.299 * r + 0.587 * g + 0.114 * b


def is_overlay_pixel(r: int, g: int, b: int) -> bool:
    if max(r, g, b) - min(r, g, b) > 55:
        return False
    lum = luminance(r, g, b)
    if lum < 155:
        return False
    if r > 145 and g > 135 and b > 115 and lum < 245:
        return True
    return False


def main() -> None:
    if not SRC.exists():
        raise SystemExit(f"Missing {SRC}")

    if not BAK.exists():
        shutil.copy2(SRC, BAK)

    img = Image.open(BAK).convert("RGB")
    px = img.load()
    w, h = img.size
    mask = [[False] * w for _ in range(h)]

    for y in range(h):
        for x in range(w):
            if is_overlay_pixel(*px[x, y]):
                mask[y][x] = True

    dilated = [[False] * w for _ in range(h)]
    for y in range(h):
        for x in range(w):
            if not mask[y][x]:
                continue
            for dy in range(-2, 3):
                for dx in range(-2, 3):
                    ny, nx = y + dy, x + dx
                    if 0 <= ny < h and 0 <= nx < w:
                        dilated[ny][nx] = True

    changed = 0
    for y in range(h):
        for x in range(w):
            if not dilated[y][x]:
                continue
            samples = []
            for dy in range(-4, 5):
                for dx in range(-4, 5):
                    if dx == 0 and dy == 0:
                        continue
                    ny, nx = y + dy, x + dx
                    if 0 <= ny < h and 0 <= nx < w and not dilated[ny][nx]:
                        samples.append(px[nx, ny])
            if samples:
                r = sum(s[0] for s in samples) // len(samples)
                g = sum(s[1] for s in samples) // len(samples)
                b = sum(s[2] for s in samples) // len(samples)
                px[x, y] = (r, g, b)
                changed += 1

    img = img.filter(ImageFilter.GaussianBlur(radius=0.35))
    img.save(OUT, optimize=True)
    print(f"Cleaned {changed} overlay pixels in {OUT}")


if __name__ == "__main__":
    main()
